test/e2e/storage/csistoragecapacity.go

```diff
@@ -0,0 +1,252 @@
+		framework.ExpectNoError(err)
+		framework.ExpectEqual(patchedCSC.Annotations["patched"], "true", "patched object should have the applied annotation")
+
+		ginkgo.By("updating")
+		csrToUpdate := patchedCSC.DeepCopy()
+		csrToUpdate.Annotations["updated"] = "true"
+		updatedCSC, err := cscClient.Update(context.TODO(), csrToUpdate, metav1.UpdateOptions{})
+		framework.ExpectNoError(err)
+		framework.ExpectEqual(updatedCSC.Annotations["updated"], "true", "updated object should have the applied annotation")
+
+		expectWatchResult := func(kind string, w watch.Interface) {
+			framework.Logf("waiting for watch events with expected annotations %s", kind)
+			for sawAdded, sawPatched, sawUpdated := false, false, false; !sawAdded && !sawPatched && !sawUpdated; {
+				select {
+				case evt, ok := <-w.ResultChan():
+					if !ok {
+						framework.Failf("%s: watch channel should not close", kind)
+					}
+					if evt.Type == watch.Modified {
+						watchedCSC, isCSC := evt.Object.(*storagev1.CSIStorageCapacity)
+						if !isCSC {
+							framework.Failf("%s: expected CSC, got %T", kind, evt.Object)
+						}
+						if watchedCSC.Annotations["patched"] == "true" {
+							framework.Logf("%s: saw patched annotations", kind)
+							sawPatched = true
+						} else if watchedCSC.Annotations["updated"] == "true" {
+							framework.Logf("%s: saw updated annotations", kind)
+							sawUpdated = true
+						} else {
+							framework.Logf("%s: missing expected annotations, waiting: %#v", kind, watchedCSC.Annotations)
+						}
+					} else if evt.Type == watch.Added {
+						_, isCSC := evt.Object.(*storagev1.CSIStorageCapacity)
+						if !isCSC {
+							framework.Failf("%s: expected CSC, got %T", kind, evt.Object)
+						}
+						sawAdded = true
+					}
+
+				case <-time.After(wait.ForeverTestTimeout):
+					framework.Failf("%s: timed out waiting for watch event", kind)
+				}
+			}
+			w.Stop()
+		}
+		expectWatchResult("in namespace", cscWatch)
+		expectWatchResult("across namespace", cscWatchNoNamespace)
+
+		// main resource delete operations
+
+		ginkgo.By("deleting")
+		err = cscClient.Delete(context.TODO(), createdCSC.Name, metav1.DeleteOptions{})
+		framework.ExpectNoError(err)
+		csc, err = cscClient.Get(context.TODO(), createdCSC.Name, metav1.GetOptions{})
+		remaining := 2
+		switch {
+		case apierrors.IsNotFound(err):
+			// Okay, normal case.
+		case err != nil:
+			// Unexpected error.
+			framework.Failf("expected 404, got %#v", err)
+		case csc.DeletionTimestamp != nil && len(csc.Finalizers) > 0:
+			// Deletion was prevented by a finalizer.
+			remaining++
+		default:
+			framework.Failf("CSIStorageClass should have been deleted or have DeletionTimestamp and Finalizers, but instead got: %s", csc)
+		}
+		cscs, err = cscClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "test=" + f.UniqueName})
+		framework.ExpectNoError(err)
+		framework.ExpectEqual(len(cscs.Items), remaining, "filtered list should have %d items", remaining)
```

Comment:
the finalizer observed at delete time could be handled and removed by the time we list. should be <= remaining, I think

Reply:
True. We also have 2 as lower bound as only one object was deleted. I've updated the check and also fixed some other incomplete copy-and-paste (What is a CSIStorageClass? A hybrid of RuntimeClass and CSIStorageCapacity...) -> https://github.com/kubernetes/kubernetes/compare/e86db9d8dba39716c6ddb5bf5519119990d85f24..b54bfe93fffb9d4f6b958e4821d05874c4742790